Hardcoded blob.core.windows.net makes azurerm not work in China (proably Germany and Government environments too)

### Terraform Version
Terraform v0.8.4-dev (6bafd4c8961f13edf0effda9ebfeb208c5ec5ffe+CHANGES)

(modified version with patch published in #11131)

### Affected Resource(s)
resource_azure_instance
resource_azure_data_disk

### Terraform Configuration Files
https://www.dropbox.com/s/ij7u77i37mvwbb1/invalid_character_error.zip?dl=0
(same as  #11131)

### Debug Output
```
2017/01/11 14:09:10 [DEBUG] apply: azurerm_storage_container.vhds: executing Apply
azurerm_storage_container.vhds: Creating...
  container_access_type: "" => "private"
  name:                  "" => "vhds"
  properties.%:          "" => "<computed>"
  resource_group_name:   "" => "d8kp5"
  storage_account_name:  "" => "curoversed8kp5"
2017/01/11 14:09:10 [DEBUG] plugin: terraform: azurerm-provider (internal) 2017/01/11 14:09:10 [DEBUG] AzureRM Request:
2017/01/11 14:09:10 [DEBUG] plugin: terraform: POST /subscriptions/c1d7e503-faf6-4a7e-829a-bc27898c09ac/resourceGroups/d8kp5/providers/Microsoft.Storage/storageAccounts/curoversed8kp5/listKeys?api-version=2016-01-01 HTTP/1.1
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Host: management.chinacloudapi.cn
2017/01/11 14:09:10 [DEBUG] plugin: terraform: User-Agent: HashiCorp-Terraform-v0.8.4-dev
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Content-Length: 0
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Authorization: Bearer e..w
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Accept-Encoding: gzip
2017/01/11 14:09:10 [DEBUG] plugin: terraform: 
2017/01/11 14:09:10 [DEBUG] plugin: terraform: 
2017/01/11 14:09:10 [DEBUG] plugin: terraform: azurerm-provider (internal) 2017/01/11 14:09:10 [DEBUG] AzureRM Response for https://management.chinacloudapi.cn/subscriptions/c..c/resourceGroups/d8kp5/providers/Microsoft.Storage/storageAccounts/curoversed8kp5/listKeys?api-version=2016-01-01:
2017/01/11 14:09:10 [DEBUG] plugin: terraform: HTTP/1.1 200 OK
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Transfer-Encoding: chunked
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Cache-Control: no-cache
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Content-Type: application/json
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Date: Wed, 11 Jan 2017 14:09:10 GMT
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Expires: -1
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Pragma: no-cache
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Server: Microsoft-Azure-Storage-Resource-Provider/1.0
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Server: Microsoft-HTTPAPI/2.0
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Strict-Transport-Security: max-age=31536000; includeSubDomains
2017/01/11 14:09:10 [DEBUG] plugin: terraform: Vary: Accept-Encoding
2017/01/11 14:09:10 [DEBUG] plugin: terraform: X-Ms-Correlation-Request-Id: 2407d66b-7331-49d7-8b00-7e7027992956
2017/01/11 14:09:10 [DEBUG] plugin: terraform: X-Ms-Ratelimit-Remaining-Subscription-Writes: 1196
2017/01/11 14:09:10 [DEBUG] plugin: terraform: X-Ms-Request-Id: 2407..956
2017/01/11 14:09:10 [DEBUG] plugin: terraform: X-Ms-Routing-Request-Id: CHINAEAST:20170111T140910Z:240..956
2017/01/11 14:09:10 [DEBUG] plugin: terraform: 
2017/01/11 14:09:10 [DEBUG] plugin: terraform: 121
2017/01/11 14:09:10 [DEBUG] plugin: terraform: {"keys":...}
2017/01/11 14:09:10 [DEBUG] plugin: terraform: 
2017/01/11 14:09:10 [DEBUG] plugin: terraform: 0
2017/01/11 14:09:10 [DEBUG] plugin: terraform: 
2017/01/11 14:09:10 [DEBUG] plugin: terraform: 
2017/01/11 14:09:10 [DEBUG] plugin: terraform: azurerm-provider (internal) 2017/01/11 14:09:10 [INFO] Creating container "vhds" in storage account "curoversed8kp5".
2017/01/11 14:09:10 [DEBUG] root: eval: *terraform.EvalWriteState
2017/01/11 14:09:10 [DEBUG] root: eval: *terraform.EvalApplyProvisioners
2017/01/11 14:09:10 [DEBUG] root: eval: *terraform.EvalIf
2017/01/11 14:09:10 [DEBUG] root: eval: *terraform.EvalWriteState
2017/01/11 14:09:10 [DEBUG] root: eval: *terraform.EvalWriteDiff
2017/01/11 14:09:10 [DEBUG] root: eval: *terraform.EvalApplyPost
2017/01/11 14:09:10 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* azurerm_storage_container.vhds: Error creating container "vhds" in storage account "curoversed8kp5": Put https://curoversed8kp5.blob.core.windows.net/vhds?restype=container: dial tcp: lookup curoversed8kp5.blob.core.windows.net on 172.16.0.23:53: no such host
2017/01/11 14:09:10 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* azurerm_storage_container.vhds: Error creating container "vhds" in storage account "curoversed8kp5": Put https://curoversed8kp5.blob.core.windows.net/vhds?restype=container: dial tcp: lookup curoversed8kp5.blob.core.windows.net on 172.16.0.23:53: no such host


Error applying plan:

1 error(s) occurred:

* azurerm_storage_container.vhds: Error creating container "vhds" in storage account "curoversed8kp5": Put https://curoversed8kp5.blob.core.windows.net/vhds?restype=container: dial tcp: lookup curoversed8kp5.blob.core.windows.net on 172.16.0.23:53: no such host

```

### Expected Behavior
create a disk in the corresponding endpoint for china

### Actual Behavior

how is azurerm_storage_account.storage.primary_blob_endpoint being used? I'm getting blob.core.windows.net But actually is blob.core.chinacloudapi.cn:

azure account env show AzureChinaCloud
(..)
data:    storageEndpointSuffix:                             .core.chinacloudapi.cn

### Steps to Reproduce
terraform apply -state mystate.tfstate -var-file invalid_character_error.tfvars -target=azurerm_virtual_machine.api

### Important Factoids

```
builtin/providers/azurerm/resource_arm_virtual_machine.go:      // VHD URI is in the form: https://storageAccountName.blob.core.windows.net/containerName/blobName
builtin/providers/azure/resource_azure_instance.go:     osDiskBlobStorageURL  = "http://%s.blob.core.windows.net/" + storageContainterName + "/" + osDiskBlobNameFormat
builtin/providers/azure/resource_azure_data_disk.go:const dataDiskBlobStorageURL = "http://%s.blob.core.windows.net/disks/%s.vhd"

```

--- a/builtin/providers/azurerm/config.go
+++ b/builtin/providers/azurerm/config.go
@@ -32,6 +32,7 @@ type ArmClient struct {
 	clientId       string
 	tenantId       string
 	subscriptionId string
+	environment    azure.Environment
 
 	StopContext context.Context
 
@@ -131,13 +132,6 @@ func setUserAgent(client *autorest.Client) {
 // getArmClient is a helper method which returns a fully instantiated
 // *ArmClient based on the Config's current settings.
 func (c *Config) getArmClient() (*ArmClient, error) {
-	// client declarations:
-	client := ArmClient{
-		clientId:       c.ClientID,
-		tenantId:       c.TenantID,
-		subscriptionId: c.SubscriptionID,
-	}
-
 	// detect cloud from environment
 	env, envErr := azure.EnvironmentFromName(c.Environment)
 	if envErr != nil {
@@ -147,6 +141,14 @@ func (c *Config) getArmClient() (*ArmClient, error) {
 		if env, innerErr = azure.EnvironmentFromName(wrapped); innerErr != nil {
 			return nil, envErr
 		}
+	}
+
+	// client declarations:
+	client := ArmClient{
+		clientId:       c.ClientID,
+		tenantId:       c.TenantID,
+		subscriptionId: c.SubscriptionID,
+		environment:    env,
 	}
 
 	rivieraClient, err := riviera.NewClient(&riviera.AzureResourceManagerCredentials{
@@ -470,7 +472,8 @@ func (armClient *ArmClient) getBlobStorageClientForStorageAccount(resourceGroupN
 		return nil, false, nil
 	}
 
-	storageClient, err := mainStorage.NewBasicClient(storageAccountName, key)
+	storageClient, err := mainStorage.NewClient(storageAccountName, key, armClient.environment.StorageEndpointSuffix,
+		mainStorage.DefaultAPIVersion, true)
 	if err != nil {
 		return nil, true, fmt.Errorf("Error creating storage client for storage account %q: %s", storageAccountName, err)
 	}
@@ -488,7 +491,8 @@ func (armClient *ArmClient) getFileServiceClientForStorageAccount(resourceGroupN
 		return nil, false, nil
 	}
 
-	storageClient, err := mainStorage.NewBasicClient(storageAccountName, key)
+	storageClient, err := mainStorage.NewClient(storageAccountName, key, armClient.environment.StorageEndpointSuffix,
+		mainStorage.DefaultAPIVersion, true)
 	if err != nil {
 		return nil, true, fmt.Errorf("Error creating storage client for storage account %q: %s", storageAccountName, err)
 	}
@@ -506,7 +510,8 @@ func (armClient *ArmClient) getTableServiceClientForStorageAccount(resourceGroup
 		return nil, false, nil
 	}
 
-	storageClient, err := mainStorage.NewBasicClient(storageAccountName, key)
+	storageClient, err := mainStorage.NewClient(storageAccountName, key, armClient.environment.StorageEndpointSuffix,
+		mainStorage.DefaultAPIVersion, true)
 	if err != nil {
 		return nil, true, fmt.Errorf("Error creating storage client for storage account %q: %s", storageAccountName, err)
 	}
@@ -524,7 +529,8 @@ func (armClient *ArmClient) getQueueServiceClientForStorageAccount(resourceGroup
 		return nil, false, nil
 	}
 
-	storageClient, err := mainStorage.NewBasicClient(storageAccountName, key)
+	storageClient, err := mainStorage.NewClient(storageAccountName, key, armClient.environment.StorageEndpointSuffix,
+		mainStorage.DefaultAPIVersion, true)
 	if err != nil {
 		return nil, true, fmt.Errorf("Error creating storage client for storage account %q: %s", storageAccountName, err)
 	}

